pages/docs/manual/latest/polymorphic-variant.mdx

```diff
@@ -0,0 +1,273 @@
+
+```res
+type users = [ #admin | #moderator | #user ]
+
+let admin = #admin
+```
+
+In rare cases (mostly for JS interop reasons), it's also possible to define invalid identifiers, such as hypens or numbers:
+
+```res
+type numbers = [#\"1" | #\"2"]
+let one = #\"1"
+```
+
+### Constructor Arguments
+
+This is equivalent to what we've already learned with common variants:
+
+```res
+type account = [
+  | #Anonymous
+  | #Instagram(string)
+  | #Facebook(string, int)
+]
+
+let acc: account = #Instagram("test")
+```
+
+### Annotations with Upper / Lower Bound Constraints
+
+There's also a way to define an "upper" and "lower" bound constraint for a poly variant type (that's why they are called _Polymorphic Variants_). Here is what it looks like in type annotations:
+
+```res
+// Only #Red allowed, no upper / lower bound (= exact)
+let basic: [#Red] = #Red
+
+// May contain #Red, or any other value (open variant)
+// here, foreground will be an inferred type [> #Red | #Green]
+let foreground: [> #Red] = #Green
+
+// The value must be "one of" #Red | #Blue
+// Only #Red and #Blue are valid values
+let background: [< #Red | #Blue] = #Red
+```
+
+Don't worry about the upper / lower bound feature (aka polymorphism) just yet, since this is a very advanced topic that's often not really needed. For the sake of completeness, we mention a few details about it [later on](#lower--upper-bound-constraints).
+
+
+## Polymorphic Variants are Structurally Typed
+
+As we've already seen in the section above, poly variants are treated a little bit differently than common variants. Most notably, we don't need any explicit type definition to define a value.
+
+```res
+// inferred as [> #Red]
+let color = #Red
+```
+
+The compiler will automatically infer the `color` binding as a value of type `[> #Red]`, which means `color` will type check with any other poly variant type that defines `#Red` in its constructors.
+
+This means that you can essentially mix and match poly variant values from different sources, as long as all constructors are defined in the final interface. For example:
+
+```res
+type rgb = [#Red | #Green | #Blue]
+
+let colors: array<rgb> = [#Red]
+
+// `other` is inferred as a type of array<[> #Green]>
+let other = [#Green]
+
+// Because `other` is of type `array<[> Green]>`,
+// this will type check even though we didn't define
+// `other`to be of type rgb
+let all = Belt.Array.concat(colors, other)
+```
+
+As you can see in the example above, the type checker doesn't really care that `color` is not annotated as an `array<rgb>` type. As soon as it hits the first constraint (`Belt.Array.concat`), it will try to check if `colors` and `other` unify into one polymorphic type. If there's a mismatch, you will get an error on the `Belt.Array.concat` call.
+
+**That means that it is very easy to get confusing type errors on the wrong locations!**
+
+For instance, if I'd make a typo like this:
+
+```res
+let other = [#GreeN]
```

Comment:
```suggestion
let other = [#Green]
```

Reply:
hehe, this is the actual trypo I tried to demonstrate: `For instance, if I'd make a typo like this:`

Maybe worth adding a comment on top

Reply:
Oops, sorry, I missed the introductory line before. 😅